proposals/2023/03-15_worker-auth.md

```diff
@@ -0,0 +1,177 @@
+- Add a `GET` validate token endpoint. One of the drawbacks of using HMAC token signing is the fact that there is no way to check validity of any given token without it being parsed by the entity with the private key (in this case, the server). By adding a simple endpoint that validates a token came from the server the worker expected, we will be able to verify tokens from the server to the worker (e.g. canceling a build).
+
+### Worker Design
+
+The worker design is more tricky, but will look something like this:
+
+1. On start up, the worker will enter a dead loop, where it will do nothing (no pulling from queue, checking in, etc) except serve the worker API. It will then be the job of the platform administrators to "register" the worker by hitting the worker's self registration endpoint.
+2. Once the self-registration endpoint is hit, the worker will use the token gathered from the header of that incoming request and break its dead loop, thereby entering its check in loop and exec loop.
+3-a. During the check in process, the worker will exchange its registration token for an auth token, which it will then refresh on the check-in interval. The success status of the check in will be a new struct field for the worker.
+3-b. Based on the success status, the worker will proceed with its exec loop, checking before each pull whether the worker is truly verified with the server.
+4. When a build is pulled off the queue, the worker will request a build token using its auth token.
+5. In a scenario where a worker loses connection for a certain time period that renders the auth token expired, and admin is require to re-register the worker
+
+Other considerations to further improve security:
+- specification of IP address in the claims to allow restriction by IP
+- stricter expiration logic
+
+In the case where the admins would prefer to continue using the symmetric token, all the above worker design logic will use the `VELA_SERVER_SECRET` token, which will not expire and not hinder executors' ability to pull from the queue.
+
+### CLI Design
+
+Since the registration token is so quick to expire, it is important to automate the job of onboarding workers for platform admins. This means that a new CLI command will be added that will use platform admin credentials to retrieve a registration token and hit the worker self-register endpoint. This will be vital for use in automation tools.
+
+### Other things of note
+
+- The server secret would be dropped from the container environment altogether if platform admins opt in to this enhancement.
+- All auth tokens (check in + registration) have a strict expiration.
+- Registration tokens can only be generated by a platform administrator
+- Registration tokens should expire very quickly
```

Comment:
Any suggestions about how quickly? Would this be configurable?

Reply:
The default will be `1m` since the accompanying CLI changes will be quick enough to not have that be a problem. But yes it will be configurable.